### PR TITLE
feat: implement paymaster

### PR DESCRIPTION
**Pull request by yutianwu**
### Description

This pr will implements BEP https://github.com/bnb-chain/BEPs/pull/362.

### Rationale

Implement paymaster.

### Example

n/a

### Changes

Notable changes: 
* add `MsgSetBucketFlowRateLimit`

### Potential Impacts
* add new message
* affect all the fee-charging related messages of storage module, like `CreatObject`, `CreateBucket` and etc.

### CODE REVIEW

File: app/upgrade.go [hunk 1]
```diff
@@ -205,6 +205,7 @@ func (app *App) registerPawneeUpgradeHandler() {
 			app.Logger().Info("upgrade to ", plan.Name)
 			app.GashubKeeper.SetMsgGasParams(ctx, *gashubtypes.NewMsgGasParamsWithFixedGas(sdk.MsgTypeURL(&storagemoduletypes.MsgUpdateObjectContent{}), 1.2e3))
 			app.GashubKeeper.SetMsgGasParams(ctx, *gashubtypes.NewMsgGasParamsWithFixedGas(sdk.MsgTypeURL(&storagemoduletypes.MsgCancelUpdateObjectContent{}), 1.2e3))
+			app.GashubKeeper.SetMsgGasParams(ctx, *gashubtypes.NewMsgGasParamsWithFixedGas(sdk.MsgTypeURL(&storagemoduletypes.MsgSetBucketFlowRateLimit{}), 1.2e3))
```
Comment (zjubfd) [inline app/upgrade.go:208]: please change to new hardfork.

File: proto/greenfield/storage/events.proto [hunk 2]
```diff
@@ -627,3 +627,18 @@ message EventCancelUpdateObjectContent {
     (gogoproto.nullable) = false
   ];
 }
+
+message EventSetBucketFlowRateLimit {
```
Comment (zjubfd) [inline proto/greenfield/storage/events.proto:631]: Does this need any change on greenfieldscan, please talk to downstream service dev.

File: x/storage/keeper/keeper_rate_limit.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+func (k Keeper) SetBucketFlowRateLimit(ctx sdk.Context, operator sdk.AccAddress, bucketOwner sdk.AccAddress, paymentAccount sdk.AccAddress, bucketName string, rateLimit sdkmath.Int) error {
+	// check the operator is the same as the payment account owner
+	if !k.paymentKeeper.IsPaymentAccountOwner(ctx, paymentAccount, operator) {
+		return paymenttypes.ErrNotPaymentAccountOwner
+	}
+
+	// get the bucket
+	bucket, found := k.GetBucketInfo(ctx, bucketName)
+	if !found {
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:22]: we do allow `SetBucketFlowRateLimit` before the bucket is created, dont we?

File: x/storage/keeper/keeper_rate_limit.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+		return types.ErrNoSuchBucket
+	}
+
+	// check the bucket owner is the same as the bucket owner
+	if bucket.Owner != bucketOwner.String() {
+		return types.ErrInvalidBucketOwner
+	}
+
+	// if the bucket does not use the payment account, just set the flow rate limit
+	if bucket.PaymentAddress != paymentAccount.String() {
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:32]: such check is not necessary

File: x/storage/keeper/keeper_rate_limit.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+		}
+	} else {
+		err := k.setNonZeroBucketFlowRateLimit(ctx, bucket, paymentAccount, bucketName, rateLimit)
+		if err != nil {
+			return err
+		}
+	}
+
+	// set the flow rate limit to the store
+	k.setBucketFlowRateLimit(ctx, paymentAccount, bucketName, &types.BucketFlowRateLimit{
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:54]: I think `bucketOwner` should be one of the key?  Otherwise someone can back run this tx, create a same `bucketName` before the expected create bucket tx come in.

File: x/storage/keeper/msg_server.go [hunk 6]
```diff
@@ -755,3 +755,27 @@ func (k Keeper) verifyGVGSignatures(ctx sdk.Context, bucketID math.Uint, dstSP *
...
+	paymentAccountAddr := sdk.MustAccAddressFromHex(msg.PaymentAddress)
+
+	err := k.Keeper.SetBucketFlowRateLimit(ctx, operatorAddr, bucketOwnerAddr, paymentAccountAddr, msg.BucketName, msg.FlowRateLimit)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = ctx.EventManager().EmitTypedEvents(&types.EventSetBucketFlowRateLimit{
+		Operator:       operatorAddr.String(),
+		BucketName:     msg.BucketName,
```
Comment (zjubfd) [inline x/storage/keeper/msg_server.go:773]: I will suggest add the `bucketOwner` as well.

File: x/storage/keeper/keeper_rate_limit.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+		// set the flow rate limit to the store
+		k.setBucketFlowRateLimit(ctx, paymentAccount, bucketName, &types.BucketFlowRateLimit{
+			FlowRateLimit: rateLimit,
+		})
+		return nil
+	}
+
+	// below are the logic for the bucket using the payment account
+	if rateLimit.IsZero() {
+		err := k.setZeroBucketFlowRateLimit(ctx, bucket, paymentAccount, bucketName)
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:42]: after this, the object inside the bucket may not be readable any longer, dont we need to set status of the bucket?

File: x/storage/keeper/keeper_rate_limit.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+		if err != nil {
+			return err
+		}
+	} else {
+		err := k.setNonZeroBucketFlowRateLimit(ctx, bucket, paymentAccount, bucketName, rateLimit)
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:47]: I will suggest we combine `setZeroBucketFlowRateLimit` and `setNonZeroBucketFlowRateLimit` into one function, because if the ratelimit is small, `setNonZeroBucketFlowRateLimit` will cause the bucket froze as well?

File: x/storage/keeper/keeper_rate_limit.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+		bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+		if err != nil {
+			return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+		}
+		totalOutFlowRate := getTotalOutFlowRate(bill.Flows)
+		if totalOutFlowRate.GT(rateLimit) {
+			return fmt.Errorf("the total out flow rate(%s) of the bucket is greater than the new rate limit(%s)", totalOutFlowRate.String(), rateLimit.String())
+		}
+
+		return k.chargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:134]: shall we set the status of the bucket?

File: x/storage/keeper/keeper_rate_limit.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	// if the rate limit is not set t0 zero before, we should check the net flow rate of the bucket
+	bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+	if err != nil {
+		return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+	}
+	totalOutFlowRate := getTotalOutFlowRate(bill.Flows)
+	if totalOutFlowRate.GT(rateLimit) {
+		return fmt.Errorf("the total out flow rate(%s) of the bucket is greater than the new rate limit(%s)", totalOutFlowRate.String(), rateLimit.String())
```
Comment (zjubfd) [inline x/storage/keeper/keeper_rate_limit.go:144]: it is a bit not resonable  to not allow to set a ratelimit that below the current usage.

File: proto/greenfield/storage/common.proto [hunk 11]
```diff
@@ -110,3 +110,12 @@ message LocalVirtualGroup {
   // Notice that the minimum unit of charge is 128K
   uint64 total_charge_size = 4;
 }
+
+message BucketFlowRateLimit {
```
Comment (forcodedancing) [inline proto/greenfield/storage/common.proto:114]: `BucketFlowRateLimit` is just `Int`. Why we need this instead of using `Int`?

File: proto/greenfield/storage/tx.proto [hunk 12]
```diff
@@ -679,3 +681,24 @@ message MsgCancelUpdateObjectContent {
...
+
+message MsgSetBucketFlowRateLimit {
+  option (cosmos.msg.v1.signer) = "operator";
+
+  // operator defines the account address of the operator, either the object owner or the updater with granted permission.
+  string operator = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // bucket_name defines the name of the bucket
+  string bucket_name = 2;
+  // bucket_owner defines the account address of the bucket owner
+  string bucket_owner = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```
Comment (forcodedancing) [inline proto/greenfield/storage/tx.proto:693]: `payment_address defines an account address specified by bucket owner` - is this wrong?

File: x/storage/keeper/keeper_rate_limit.go [hunk 13]
```diff
@@ -0,0 +1,232 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	paymenttypes "github.com/bnb-chain/greenfield/x/payment/types"
+	"github.com/bnb-chain/greenfield/x/storage/types"
```
Comment (forcodedancing) [inline x/storage/keeper/keeper_rate_limit.go:10]: Duplicated imports here.

File: x/storage/types/crosschain_test.go [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package types
+
+import (
+	"encoding/hex"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func TestDecoe(t *testing.T) {
```
Comment (forcodedancing) [inline x/storage/types/crosschain_test.go:10]: typo. Is this file needed?

File: x/storage/types/message_set_bucket_rate_limit.go [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+import (
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/bnb-chain/greenfield/types/s3util"
+)
+
+const TypeMsgSetBucketFlowRateLimit = "cancel_migrate_bucket"
```
Comment (forcodedancing) [inline x/storage/types/message_set_bucket_rate_limit.go:11]: `cancel_migrate_bucket` is not correct.

File: x/storage/keeper/keeper_rate_limit.go [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+func (k Keeper) isBucketFlowRateLimitSetToZero(ctx sdk.Context, paymentAccount sdk.AccAddress, bucketName string) bool {
+	rateLimit, found := k.getBucketFlowRateLimit(ctx, paymentAccount, bucketName)
+	if !found {
+		return false
+	}
+	return rateLimit.FlowRateLimit.IsZero()
+}
+
+// isBucketFlowRateUnderLimit checks if the flow rate of the bucket is under the flow rate limit
+func (k Keeper) isBucketFlowRateUnderLimit(ctx sdk.Context, paymentAccount, bucketOwner sdk.AccAddress, bucketName string, userFlows paymenttypes.UserFlows) error {
```
Comment (forcodedancing) [inline x/storage/keeper/keeper_rate_limit.go:182]: This function can call `isBucketFlowRateUnderLimitWithRate` to reduce duplicated codes.

File: x/storage/keeper/keeper_rate_limit.go [hunk 17]
```diff
@@ -0,0 +1,232 @@
+package keeper
```
Comment (forcodedancing) [inline x/storage/keeper/keeper_rate_limit.go:1]: Can we rename this file? Other files do not have `keeper_` in their names.

File: x/storage/keeper/payment.go [hunk 18]
```diff
@@ -129,6 +145,25 @@ func (k Keeper) lockObjectStoreFee(ctx sdk.Context, bucketInfo *storagetypes.Buc
 		})
 	}
 
+	if ctx.IsUpgraded(upgradetypes.Pawnee) {
+		internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+		internalBucketInfo.PriceTime = timestamp
+		nextBill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
```
Comment (forcodedancing) [inline x/storage/keeper/payment.go:151]: Most buckets do not have rate limit. Can we check if there is rate limiter earlier and break quickly? Otherwise, it seems it will affect the performance for the general cases.

File: x/storage/keeper/payment.go [hunk 19]
```diff
@@ -309,12 +346,35 @@ func (k Keeper) ChargeViaBucketChange(ctx sdk.Context, bucketInfo *storagetypes.
 		return fmt.Errorf("get new bucket bill failed: %s %w", bucketInfo.BucketName, err)
 	}
 
-	// charge according to bill change
-	err = k.ApplyBillChanges(ctx, prevBill, newBill)
-	if err != nil {
-		ctx.Logger().Error("charge via bucket change failed", "bucket", bucketInfo.BucketName, "err", err.Error())
-		return err
+	if ctx.IsUpgraded(upgradetypes.Pawnee) {
```
Comment (forcodedancing) [inline x/storage/keeper/payment.go:349]: Can we refactor the code a bit? 
1. move line 336 to the inner of `if ctx.IsUpgraded(upgradetypes.Pawnee)`
2. in the `if ctx.IsUpgraded(upgradetypes.Pawnee)` we just set the value of `isPreviousBucketSuspended`. And out of the `if` we can take different actions based on `isPreviousBucketSuspended`.

File: x/storage/keeper/payment.go [hunk 20]
```diff
@@ -366,6 +426,19 @@ func (k Keeper) ChargeViaObjectChange(ctx sdk.Context, bucketInfo *storagetypes.
 
 	userFlows.Flows = append(userFlows.Flows, getNegFlows(preOutFlows)...)
 	userFlows.Flows = append(userFlows.Flows, newOutFlows...)
+
+	if ctx.IsUpgraded(upgradetypes.Pawnee) {
```
Comment (forcodedancing) [inline x/storage/keeper/payment.go:430]: Most buckets do not have rate limit. Can we check if there is rate limiter earlier and break quickly? Otherwise, it seems it will affect the performance for the general cases.

File: proto/greenfield/storage/common.proto (hunk 11 above)
Comment (yutianwu) [reply proto/greenfield/storage/common.proto:114]: we can add another fields

File: x/storage/keeper/payment.go (hunk 18 above)
Comment (yutianwu) [reply x/storage/keeper/payment.go:151]: sure

File: x/storage/keeper/payment.go (hunk 20 above)
Comment (yutianwu) [reply x/storage/keeper/payment.go]: there is no heavy operations under this function, i think it's fine here

File: x/storage/keeper/grpc_query.go [hunk 21]
```diff
@@ -62,8 +62,13 @@ func (k Keeper) HeadBucket(goCtx context.Context, req *types.QueryHeadBucketRequ
 
 	bucketInfo, found := k.GetBucketInfo(ctx, req.BucketName)
 	if found {
+		extraInfo, err := k.GetBucketExtraInfo(ctx, req.BucketName)
```
[the remark this answers is not in the document]
Comment (yutianwu) [reply x/storage/keeper/grpc_query.go]: sure

File: proto/greenfield/storage/tx.proto (hunk 12 above)
Comment (yutianwu) [reply proto/greenfield/storage/tx.proto:693]: fixed

File: x/storage/keeper/bucket_rate_limit.go [hunk 22]
```diff
@@ -0,0 +1,310 @@
...
+	if !k.paymentKeeper.IsPaymentAccountOwner(ctx, paymentAccount, operator) {
+		return paymenttypes.ErrNotPaymentAccountOwner
+	}
+
+	// get the bucket
+	bucket, found := k.GetBucketInfo(ctx, bucketName)
+
+	// check the bucket owner is the same as the bucket owner of the bucket
+	if found && bucket.Owner != bucketOwner.String() {
+		return fmt.Errorf("the bucket owner is not the same as the bucket owner of the bucket")
```
Comment (zjubfd) [inline x/storage/keeper/bucket_rate_limit.go:24]: I think it is ok, dont have to check `bucket.Owner != bucketOwner.String()`, because we can not guarante this if the bucket is created later.

File: x/storage/keeper/bucket_rate_limit.go [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+			FlowRateLimit: rateLimit,
+			BucketOwner:   bucketOwner.String(),
+		})
+		return nil
+	}
+
+	// delete the flow rate limit for different bucket owner, since there may be a wrong rate limit or
+	// the previous rate limit set for the bucket which owned by the different bucket owner, for example,
+	// the previous bucket owner delete the bucket and the current bucket owner create a new bucket with the same name
+	k.deleteBucketFlowRateLimitForDifferentBucketOwner(ctx, paymentAccount, bucketOwner, bucketName)
```
Comment (zjubfd) [inline x/storage/keeper/bucket_rate_limit.go:40]: I strongly suggest that you use the `owner` + `bucketname` + `payment account` as the key of the ratelimit, so that you dont have to delete an already exist `FlowRateLimit`.
Because in some edge case you delete the previous ratelimit, what if there is a bucket still use it.

File: x/storage/keeper/bucket_rate_limit.go [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+
+	// set the flow rate limit for the bucket for the current bucket owner
+	err := k.setFlowRateLimit(ctx, bucket, paymentAccount, bucketName, rateLimit)
```
Comment (zjubfd) [inline x/storage/keeper/bucket_rate_limit.go:43]: `setFlowRateLimit` the function name seem not good, it is simliar to `setBucketFlowRateLimit`, but it does a lot of things.

File: x/storage/keeper/bucket_rate_limit.go [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+		totalOutFlowRate := getTotalOutFlowRate(currentBill.Flows)
+
+		// if the total net flow rate is less than or equal to the new rate limit,
+		// resume the charge of the bucket and delete the flow rate limit status
+		if totalOutFlowRate.LTE(rateLimit) {
+			err := k.chargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+			if err != nil {
+				return fmt.Errorf("charge bucket failed: %s %s", bucketInfo.BucketName, err.Error())
+			}
+			k.deleteBucketFlowRateLimitStatus(ctx, bucketName)
```
Comment (zjubfd) [inline x/storage/keeper/bucket_rate_limit.go:135]: When I bucket is deleted, seems this `FlowRateLimitStatus` is not deleted

Comment (zjubfd) [reply x/storage/keeper/bucket_rate_limit.go:135]: Also please emit event when do and undo `FlowRateLimitStatus`

File: x/storage/keeper/payment.go [hunk 26]
```diff
@@ -309,12 +346,43 @@ func (k Keeper) ChargeViaBucketChange(ctx sdk.Context, bucketInfo *storagetypes.
...
-	// charge according to bill change
-	err = k.ApplyBillChanges(ctx, prevBill, newBill)
-	if err != nil {
-		ctx.Logger().Error("charge via bucket change failed", "bucket", bucketInfo.BucketName, "err", err.Error())
-		return err
+	if ctx.IsUpgraded(upgradetypes.Serengeti) {
+		isPreviousBucketLimited := k.isBucketRateLimited(ctx, bucketInfo.BucketName)
+
+		if prevPaymentAccount == bucketInfo.PaymentAddress && isPreviousBucketLimited {
+			// todo: do we need to contraint this
```
Comment (zjubfd) [inline x/storage/keeper/payment.go:353]: yeah,ok to me.

File: x/storage/keeper/payment.go [hunk 27]
```diff
@@ -309,12 +346,43 @@ func (k Keeper) ChargeViaBucketChange(ctx sdk.Context, bucketInfo *storagetypes.
...
+		}
+
+		// check the bucket's new flow rate limit
+		err := k.isBucketFlowRateUnderLimit(ctx, sdk.MustAccAddressFromHex(bucketInfo.PaymentAddress), sdk.MustAccAddressFromHex(bucketInfo.Owner), bucketInfo.BucketName, newBill)
+		if err != nil {
+			return err
+		}
+
+		if isPreviousBucketLimited {
+			err = k.ApplyBillChanges(ctx, nil, &newBill)
```
Comment (zjubfd) [inline x/storage/keeper/payment.go:364]: This seems dead code, never will be executed

File: x/storage/keeper/payment.go [hunk 28]
```diff
@@ -486,16 +575,36 @@ func (k Keeper) ChargeBucketReadStoreFee(ctx sdk.Context, bucketInfo *storagetyp
...
+	}
+
+	var flowList []types.UserFlows
+	if prevFlows != nil {
+		flowList = append(flowList, *prevFlows)
+	}
+	if currentFlows != nil {
+		flowList = append(flowList, *currentFlows)
+	}
+
```
Comment (zjubfd) [inline x/storage/keeper/payment.go:606]: what is the motivation to change the code here?

File: x/storage/keeper/payment.go (hunk 27 above)
Comment (yutianwu) [reply x/storage/keeper/payment.go:363]: it will be executed if the bucket is rate limited before

File: app/upgrade.go [hunk 29]
```diff
@@ -206,6 +206,7 @@ func (app *App) registerPawneeUpgradeHandler() {
...
 			return app.mm.RunMigrations(ctx, app.configurator, fromVM)
```
Comment (forcodedancing) [inline app/upgrade.go:210]: Forget to remove this?

File: x/storage/client/cli/tx_set_bucket_flow_rate_limit.go [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+	"github.com/bnb-chain/greenfield/x/storage/types"
+)
+
+var _ = strconv.Itoa(0)
+
+func CmdSetBucketFlowRateLimit() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "set-bucket-flow-rate-limit [bucket-name] [payment-account] [bucket-owner] [flow-rate-limit]",
+		Short: "cancel a existing bucket migration",
```
Comment (forcodedancing) [inline x/storage/client/cli/tx_set_bucket_flow_rate_limit.go:22]: The `Short` message is wrong.

File: x/storage/keeper/bucket_rate_limit.go [hunk 31]
```diff
@@ -0,0 +1,290 @@
...
+	// check the operator is the same as the payment account owner
+	if !k.paymentKeeper.IsPaymentAccountOwner(ctx, paymentAccount, operator) {
+		return paymenttypes.ErrNotPaymentAccountOwner
+	}
+
+	// get the bucket
+	bucket, found := k.GetBucketInfo(ctx, bucketName)
+
+	// if the bucket does not use the payment account, just set the flow rate limit
+	if !found || (found && (bucket.PaymentAddress != paymentAccount.String() || bucket.Owner != bucketOwner.String())) {
```
Comment (forcodedancing) [inline x/storage/keeper/bucket_rate_limit.go:23]: Why this case `bucket.Owner != bucketOwner.String()` is fine for setting rate limit?

File: x/storage/keeper/bucket_rate_limit.go [hunk 32]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	// set the flow rate limit for the bucket for the current bucket owner
+	err := k.setFlowRateLimit(ctx, bucket, paymentAccount, bucketName, rateLimit)
+	if err != nil {
+		return err
+	}
+
+	// set the flow rate limit to the store
+	k.setBucketFlowRateLimit(ctx, paymentAccount, bucketOwner, bucketName, &types.BucketFlowRateLimit{
```
Comment (forcodedancing) [inline x/storage/keeper/bucket_rate_limit.go:38]: Do we allow mutliple rate limits (e.g., from different payment account)?
If not, we should check and prevent this.

File: x/storage/keeper/payment.go [hunk 33]
```diff
@@ -309,12 +346,42 @@ func (k Keeper) ChargeViaBucketChange(ctx sdk.Context, bucketInfo *storagetypes.
...
+				return err
+			}
+			k.deleteBucketFlowRateLimitStatus(ctx, bucketInfo.BucketName)
+		} else {
+			err = k.ApplyBillChanges(ctx, &prevBill, &newBill)
+			if err != nil {
+				ctx.Logger().Error("charge via bucket change failed", "bucket", bucketInfo.BucketName, "err", err.Error())
+				return err
+			}
+		}
```
Comment (forcodedancing) [inline x/storage/keeper/payment.go:375]:
```suggestion
		err = k.ApplyBillChanges(ctx, nil, &newBill)
		if err != nil {
			ctx.Logger().Error("charge via bucket change failed", "bucket", bucketInfo.BucketName, "err", err.Error())
			return err
		}
		if isPreviousBucketLimited {

			k.deleteBucketFlowRateLimitStatus(ctx, bucketInfo.BucketName)
		} 
```

File: x/storage/keeper/msg_server.go [hunk 34]
```diff
@@ -823,3 +823,27 @@ func (k Keeper) verifyGVGSignatures(ctx sdk.Context, bucketID math.Uint, dstSP *
...
```
Comment (zjubfd) [inline x/storage/keeper/msg_server.go:841]: Can you add the `BucketOwner` field in the event?

File: x/storage/keeper/bucket_rate_limit.go [hunk 35]
```diff
@@ -0,0 +1,290 @@
...
+		if totalOutFlowRate.LTE(rateLimit) {
+			err := k.chargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+			if err != nil {
+				return fmt.Errorf("charge bucket failed: %s %s", bucketInfo.BucketName, err.Error())
+			}
+			k.deleteBucketFlowRateLimitStatus(ctx, bucketName)
+			return nil
+		}
+
+		return nil
```
Comment (zjubfd) [inline x/storage/keeper/bucket_rate_limit.go:127]: what if it is not isRateLimited before, but `totalOutFlowRate.LTE(rateLimit)`, you do not handle this case.

**Review by forcodedancing — APPROVED** [review]

**Review by alexgao001 — APPROVED** [review]
